[Docs] Clarify default value for `allow_no_indices`

matriv opened the pull request:
Add default value to each one of the usages of `allow_no_indices`
since it differs between different APIs.

Relates to: #52534

matriv committed:
[Docs] Clarify default value for `allow_no_indices`

Add a cliarification that default value is `true`, so no error
is returned if no index is matched.

Fixes: #52534

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/Indices APIs)

elasticmachine commented:
Pinging @elastic/es-docs (>docs)

javanna commented on docs/reference/rest-api/common-parms.asciidoc:
```diff
@@ -42,6 +42,7 @@ tag::allow-no-indices[]
 the request does *not* return an error
 if a wildcard expression
 or `_all` value retrieves only missing or closed indices.
+Defaults to `true`.
```
I'm afraid that the default value here depends on the API, or is it `true` for all APIs?

matriv replied:
I tried some of them like HEAD, _mapping, _forgemerge and they behave the same.

martijnvg replied:
The default is API dependant. For example the put mapping api defaults to false.

matriv replied:
I see, should we just add a sentence that the default value depends on the API instead?

jrodewig replied:
You can add the default value where the definition is included instead:

```
include::{docdir}/rest-api/common-parms.asciidoc[tag=allow-no-indices]
+
Defaults to `true`.
```

javanna replied:
@jrodewig ++

matriv committed:
add defaults to all doc references

matriv commented:
Checked all usages and added the default value to the doc references.

I have another question though, if you check this: https://github.com/elastic/elasticsearch/pull/52635/files#diff-2e0eb1b85aa4eba2d58663601327caffL39
I removed the reference since the value of the `allow_no_indices` has not effect.
You get 404 if the the index is not there regardless of the param.

I'm wondering if we should do the same for the index exists api (`HEAD /<index-pattern>`).
What do you think?

matriv committed:
fix docs

jrodewig commented on docs/reference/docs/update-by-query.asciidoc:
```diff
@@ -158,6 +158,8 @@ or omit to search all indices.
 
 include::{docdir}/rest-api/common-parms.asciidoc[tag=analyzer]
++
+Defaults to `true`.
```
I believe this is on the wrong parameter (analyzer).

However, when I tested `allow_no_indices` on the update by query API, it had no effect on whether an error was returned. Perhaps this should be removed?

matriv replied:
`localhost:9200/index-2*/_update_by_query?conflicts=proceed&allow_no_indices=false`:
```
{
    "error": {
        "root_cause": [
            {
                "type": "index_not_found_exception",
                "reason": "no such index [index-2*]",
                "resource.type": "index_or_alias",
                "resource.id": "index-2*",
                "index_uuid": "_na_",
                "index": "index-2*"
            }
        ],
        "type": "index_not_found_exception",
        "reason": "no such index [index-2*]",
        "resource.type": "index_or_alias",
        "resource.id": "index-2*",
        "index_uuid": "_na_",
        "index": "index-2*"
    },
    "status": 404
}
```

`localhost:9200/index-2*/_update_by_query?conflicts=proceed`:
```
{
    "took": 0,
    "timed_out": false,
    "total": 0,
    "updated": 0,
    "deleted": 0,
    "batches": 0,
    "version_conflicts": 0,
    "noops": 0,
    "retries": {
        "bulk": 0,
        "search": 0
    },
    "throttled_millis": 0,
    "requests_per_second": -1.0,
    "throttled_until_millis": 0,
    "failures": []
}
```

jrodewig replied:
Sorry about that. I made a mistake in my test setup.

matriv replied:
No worries at all, I didn't see your edit in time and tried it once more myself.

jrodewig replied:
However, we should still move the default value to the allow_no_indices parm.

jrodewig commented on docs/reference/indices/alias-exists.asciidoc:
```diff
@@ -36,8 +36,6 @@ include::{docdir}/rest-api/common-parms.asciidoc[tag=index]
 ==== {api-query-parms-title}
 
-include::{docdir}/rest-api/common-parms.asciidoc[tag=allow-no-indices]
-
```
👍

matriv replied:
I'm not 100% sure. 
The mechanism is different, but the outcome the same (status 200 or 404).
If `allow_no_indices=false` and no index is matched then we get an exception at the index name resolution phase which is translated to 404.
If `allow_no_indices=true` and no index is matched then no aliases are found and this is again translated to 404.

jrodewig replied:
I think we should probably remove documentation for this parameter. It's not useful to the end user and the outcome will be the same regardless. Thanks for catching this @matriv.

matriv committed:
Address comment, add for cluster state api

matriv commented:
@jrodewig FYI, I added also this: https://github.com/elastic/elasticsearch/pull/52635/commits/9902b34090f79828165c6b6bf61821d723a83e2d#diff-72be49b35e089e3c18cadc2619d1d933R99

jrodewig reviewed (approved):
LGTM. Thanks @matriv!
